src/core/thread/context.d

```diff
@@ -0,0 +1,312 @@
+/**
+ * Every thread / fiber needs stack context data structure.
+ * thread.context decouples this info into the StackContext struct.
+ * Also, Thread / Fiber have some common functionality which is provided
+ * as a super class for both in StackContextExecutor.
+ * Finally, GlobalStackContext contains commonly used global data.
+ *
+ * Copyright: Copyright Sean Kelly 2005 - 2012.
+ * License: Distributed under the
+ *      $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+ *    (See accompanying file LICENSE)
+ * Authors:   Sean Kelly, Walter Bright, Alex Rønne Petersen, Martin Nowak
+ * Source:    $(DRUNTIMESRC core/thread/osthread.d)
+ */
+
+module core.thread.context;
+
+struct StackContext
+{
+    // bottom, top of stack
+    void* bstack, tstack;
+
+    /// Slot for the EH implementation to keep some state for each stack
+    /// (will be necessary for exception chaining, etc.). Opaque as far as
+    /// we are concerned here.
+    void* ehContext;
+
+    StackContext* within, next, prev;
+}
+
+/**
+A class that represents a thread of execution that manages a stack.
+This serves primarily as a superclass for Thread and Fiber.
+*/
+class StackContextExecutor
+{
+    //
+    // The type of routine passed on thread/fiber construction.
+    //
+    enum Call
+    {
+        NO,
+        FN,
+        DG
+    }
+
+    // Common standard data for Thread / Fiber.
+    Call                m_call = Call.NO;
+    union
+    {
+        void function() m_fn;
+        void delegate() m_dg;
+    }
+
+    // Thread / Fiber entry point.  Invokes the function or delegate passed on
+    // construction (if any).
+    final void run()
```

Comment:
note that somehow this has triggered 
```
../.dub/packages/gtk-d-3.3.1/gtk-d/src/glib/Spawn.d(204,15): Error: function `glib.Spawn.Spawn.ReadFile.run` cannot override `final` function `core.thread.context.StackContextExecutor.run`
```
in build kite, why I'm not sure.

Reply:
I can't find the source of either of these files _but_ the error most probably is something like [here](https://github.com/dlang/druntime/pull/2797/files#diff-897dcc1ea84bed3f92a8ba5e1bc6e619R1576). Note that the `private` was not there before. I put it to fix an identical error (following the example of other unittests).
So, the point is that this is _user_ code and it's not a problem of `Fiber`'s code.
But this may lead to breaking change in existing code.

Reply:
Yeah its probably a visibility/protection issue, you can ask on the forums for an explanation if you can't find a solution. (Alternately you could call it something else that won't cause a collision)

Reply:
Well, the problem is that a `final` function cannot be overriden, unless it is also `private`.
_Previously_: `run` was member of `Fiber` but it was `private` so overriding in e.g. `DerivedFiber` was ok.
_Now_: `run` is a member of `StackContextExecutor` and it can't be `private` as it has to be used by `Fiber`. Because of that, overriding it causes a problem.

I think of a couple solutions to that, of which the simplest is the one you said - chaning the name.

Reply:
Hmm, perhaps doing https://github.com/dlang/druntime/pull/2797#discussion_r324444355 and having it as the opCall in this PR would be worth doing then. Saves having to come up with a name that won't collide. Or just move run back to Fiber for the time being and do that when you address https://github.com/dlang/druntime/pull/2797#discussion_r324444355 in a future PR. Up to you.

Reply:
Ok, I'll see if there are any simple things left and do that.